### PR TITLE
[SPARK-40482][SQL] Revert `SPARK-24544 Print actual failure cause when look up function failed`

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This reverts https://github.com/apache/spark/pull/21790 because it's no longer needed. It kept the original error from Hive when Spark loads builtin functions from Hive, which no longer happens as Spark has implemented all builtin functions natively.

### Why are the changes needed?

code cleanup

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

existing tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 1]
```diff
@@ -1732,11 +1731,7 @@ class SessionCatalog(
       // The function has not been loaded to the function registry, which means
       // that the function is a persistent function (if it actually has been registered
       // in the metastore). We need to first put the function in the function registry.
-      val catalogFunction = try {
-        externalCatalog.getFunction(db, funcName)
-      } catch {
-        case _: AnalysisException => failFunctionLookup(qualifiedIdent)
-      }
+      val catalogFunction = externalCatalog.getFunction(db, funcName)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1734]: not quite related but it's obviously redundant to try-catch and re-throw the error. `externalCatalog.getFunction` can throw `NoSuchFunctionExcetion` and we should just propagate it.

**Comment by cloud-fan** [comment]
cc @caneGuy @dongjinleekr @MaxGekk

**Comment by cloud-fan** [comment]
also cc @viirya

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/NoSuchItemException.scala [hunk 2]
```diff
@@ -74,19 +74,17 @@ case class NoSuchPartitionException(
...
-    override val message: String,
-    override val cause: Option[Throwable])
-  extends AnalysisException(message, cause = cause) {
+case class NoSuchFunctionException(override val message: String)
+  extends AnalysisException(message) {
 
-  def this(db: String, func: String, cause: Option[Throwable] = None) = {
+  def this(db: String, func: String) = {
     this(s"Undefined function: '$func'. " +
         s"This function is neither a registered temporary function nor " +
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/NoSuchItemException.scala:82]: nit:
```suggestion
        "This function is neither a registered temporary function nor " +
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 3]
```diff
@@ -1588,10 +1587,9 @@ class SessionCatalog(
       TableFunctionRegistry.builtin.functionExists(name)
   }
 
-  protected[sql] def failFunctionLookup(
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1591]: Hmm, where do we set `cause`? As I see, we always call `failFunctionLookup` without `cause` (except for the test case).

**Review by viirya — APPROVED** [review]
Looks good. The `cause` seems no real usage now.

**Review by dongjoon-hyun — COMMENTED** [review]
@cloud-fan . We need a new JIRA in this case to have a traceability. Could you file a JIRA issue for this reverting?

**Comment by viirya** [comment]
Oh, that's right. SPARK-24544 is long time ago, it is better to have a new JIRA.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1591]: We used to pass the `cause` parameter before, when looking up builtin functions from Hive. The code path is gone now.

**Comment by cloud-fan** [comment]
@dongjoon-hyun thanks for creating the JIRA ticket!

**Comment by cloud-fan** [comment]
The last commit is just to remove `s` from `s"..."`. I'm merging it to master, thanks for review!

**Review by dongjoon-hyun — COMMENTED** [review]
+1, LGTM. Thank you!

**Pull request closed without merging**